File: parking_system_cpp/parking_lot.cpp
#include "parking_lot.h"
#include <iostream>

using namespace std;
using namespace chrono;

ParkingLot::ParkingLot(int floorSize) {
    for (int i = static_cast<int>(Floors::BASEMENT); i <= static_cast<int>(Floors::SECOND); i++) {
        string floorName = floorTypeToString(static_cast<Floors>(i)) + '-' + static_cast<char>('0' + i);
        Floor* floor = new Floor(floorName, floorSize);
        floors.emplace_back(floor);
    }
}

Ticket* ParkingLot::fileTicket(Vehicle* vehicle) {
    Spot* spot = nullptr;
    for (Floor* floor : floors) {
        Spot* temp = floor->getFirstVacantSpot();
        if (temp) {
            spot = temp;
            break;
        }
    }

    if (!spot) throw runtime_error("Parking full");

    string ticketId = generateUUID();
    Ticket* ticket = new Ticket(vehicle, ticketId, system_clock::now(), spot, 30);
    tickets.emplace_back(ticket);
    spot->setOccupied(vehicle);
    return ticket;
}

int ParkingLot::paymentProcess(Ticket* ticket) {
    auto durationSec = duration_cast<seconds>(system_clock::now() - ticket->entryTime).count();
    ticket->spot->setFree();
    return static_cast<int>(durationSec) * ticket->hourlyRate;
}

void ParkingLot::printStatus() const {
    for (const Floor* floor : floors) {
        cout << floor->to_string();
    }
}

const vector<Floor*>& ParkingLot::getFloors() const {
    return floors;
}

const vector<Ticket*>& ParkingLot::getTickets() const {
    return tickets;
}

ParkingLot::~ParkingLot() {
    for (Floor* floor : floors) delete floor;
    for (Ticket* ticket : tickets) delete ticket;
}

File: javasrc/parking_system/ParkingLot.java
package parking_system;

import java.util.List;
import java.util.ArrayList;
import java.util.UUID;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

class ParkingLot implements Entrance, Egress {
    private final List<Ticket> tickets;
    private final List<Floor> floors;

    public ParkingLot(int floorSize) {
        this.tickets = new ArrayList<>();
        this.floors = new ArrayList<>();
        for(FloorType type : FloorType.values()) {
            Floor floor = new Floor(type.name(), floorSize);
            floors.add(floor);
        }
    }

    public Ticket fileTicket(Vehicle vehicle) {
        Spot vacantSpot = getFirstVacantSpot();
        String ticketId = UUID.randomUUID().toString();
        Ticket tkt = new Ticket(vehicle, ticketId, LocalDateTime.now(), vacantSpot, 35);
        tickets.add(tkt);
        vacantSpot.setOccupied(vehicle);
        return tkt;
    }

    public int processPayment(Ticket tk) {
        LocalDateTime entryTime = tk.entryTime;
        LocalDateTime now = LocalDateTime.now();
        long time = ChronoUnit.SECONDS.between(entryTime, now);
        tk.spot.setFree();
        return (int) time * tk.hourlyRate;
    }

    public Spot getFirstVacantSpot() {
        for(Floor floor : floors) {
            for(Spot spot : floor.getSpots()) {
                if(spot.isFree()) {
                    return spot;
                }
            }
        }

        throw new IllegalStateException("No spot available");
    }

    public List<Floor> getFloors() {
        return new ArrayList<>(floors);
    }

    public List<Ticket> getTickets() {
        return new ArrayList<>(tickets);
    }
}